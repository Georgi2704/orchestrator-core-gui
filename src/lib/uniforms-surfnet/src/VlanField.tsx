/*
 * Copyright 2019-2020 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import { usedVlans as getUsedVlans } from "api";
import { SubscriptionsContext } from "components/subscriptionContext";
import I18n from "i18n-js";
import get from "lodash/get";
import React, { HTMLProps, Ref, useContext, useEffect, useState } from "react";
import { Override, connectField, filterDOMProps, joinName, useForm } from "uniforms";
import ApplicationContext from "utils/ApplicationContext";
import { ServicePort } from "utils/types";
import { isEmpty } from "utils/Utils";
import { inValidVlan } from "validations/UserInput";

import { getPortMode } from "./logic/SubscriptionField";

function getAllNumbersForVlanRange(vlanRange: string) {
    if (vlanRange !== "0" && inValidVlan(vlanRange)) {
        //semantically invalid so we don't validate against the already used ports
        return [];
    }

    if (vlanRange === "") {
        return [];
    }

    return numbersFromGroupedArray(
        vlanRange
            .replace(/ /g, "")
            .split(",")
            .map(sl => sl.split("-").map(Number))
    );
}

function numbersFromGroupedArray(list: number[][]) {
    return list.reduce((acc, boundaries) => {
        const max = boundaries[boundaries.length - 1];
        const min = boundaries[0];
        return acc.concat(Array.from(new Array(max - min + 1), (x, i) => min + i));
    }, []);
}

function groupedArrayFromNumbers(numbers: number[]) {
    numbers = [...numbers].sort((a, b) => a - b);

    // Group by properly incrementing numbers
    let grouped_numbers = numbers.reduce((r: number[][], n) => {
        const lastSubArray: number[] = r[r.length - 1];

        if (!lastSubArray || lastSubArray[lastSubArray.length - 1] !== n - 1) {
            r.push([]);
        }

        r[r.length - 1].push(n);

        return r;
    }, []);

    return grouped_numbers.map(l => (l[0] !== l[l.length - 1] ? [l[0], l[l.length - 1]] : [l[0]]));
}

function vlanRangeFromNumbers(list: number[]) {
    return groupedArrayFromNumbers(list)
        .map(sl => sl.join("-"))
        .join(",");
}

function getVlansInUse(vlanRange: string, usedVlans: number[][]) {
    const numbers = getAllNumbersForVlanRange(vlanRange);
    return numbers.filter(num =>
        usedVlans.some(used => (used.length > 1 ? num >= used[0] && num <= used[1] : num === used[0]))
    );
}

export type VlanFieldProps = Override<
    HTMLProps<HTMLDivElement>,
    {
        disabled: boolean;
        id: string;
        inputRef?: Ref<HTMLInputElement>;
        label: string;
        description?: string;
        name: string;
        onChange(value?: string): void;
        type?: string;
        value: string;
        error?: boolean;
        showInlineError?: boolean;
        errorMessage?: string;
    }
>;

function Vlan({
    disabled,
    id,
    inputRef,
    label,
    description,
    name,
    onChange,
    type,
    value,
    error,
    showInlineError,
    errorMessage,
    ...props
}: VlanFieldProps) {
    const { model } = useForm();
    const { products } = useContext(ApplicationContext);
    const { subscriptions } = useContext(SubscriptionsContext);
    const nameArray = joinName(null, name);
    const subscriptionIdFieldName = joinName(nameArray.slice(0, -1), "subscription_id");
    const completeListFieldName = joinName(nameArray.slice(0, -2));
    const subscriptionId = get(model, subscriptionIdFieldName);
    const completeList: ServicePort[] = get(model, completeListFieldName) || [];
    const vlansExtraInUse = completeList
        .filter(item => item.subscription_id === subscriptionId)
        .filter((_item, index) => index.toString() !== nameArray.slice(-2, -1)[0])
        .map(item => item.vlan)
        .join(",");

    const subscription = get(subscriptions, subscriptionId);
    const portMode = subscription && getPortMode(subscription, products);

    let [usedVlansInIms, setUsedVlansInIms] = useState<number[][]>([]);
    let [missingInIms, setMissingInIms] = useState(false);

    let validFormat = (vlanRange: string) => {
        // If disabled, untagged or empty we don't need syntax validation
        return disabled || value === "" || value === "0" ? vlanRange === value : !inValidVlan(vlanRange);
    };

    let validateUsedVlans = (vlanRange: string) => {
        // This gets called onBlur (so we don't validate during typing) or
        // on mount to check if an untagged port is not already used
        const allUsedVlans = groupedArrayFromNumbers(usedVlans.concat(getAllNumbersForVlanRange(vlansExtraInUse)));
        const invalidFormat = !validFormat(vlanRange);

        // Can't check if input is invalid
        let inUse: number[] = [];

        if (!invalidFormat) {
            inUse = getVlansInUse(vlanRange, allUsedVlans);
        }

        return { vlansInUse: inUse, invalidFormat: invalidFormat };
    };

    useEffect(() => {
        if (subscriptionId) {
            getUsedVlans(subscriptionId)
                .then(result => {
                    setUsedVlansInIms(result);
                    setMissingInIms(false);
                })
                .catch(e => {
                    setMissingInIms(true);
                });
        }
    }, [subscriptionId]);

    let usedVlans = numbersFromGroupedArray(usedVlansInIms);

    // Filter currently used vlans because they are probably from the current subscription
    const currentVlans = getAllNumbersForVlanRange(value);
    usedVlans = usedVlans.filter(n => !currentVlans.includes(n));

    // Don't validate if disabled (untagged shows as disable but needs validation)
    let vlansInUse: number[] = [];
    let invalidFormat = false;
    if (!disabled) {
        const result = validateUsedVlans(value);
        vlansInUse = result.vlansInUse;
        invalidFormat = result.invalidFormat;
    }

    const allUsedVlans = vlanRangeFromNumbers(usedVlans.concat(getAllNumbersForVlanRange(vlansExtraInUse)));
    const placeholder = subscriptionId
        ? I18n.t("forms.widgets.vlan.placeholder")
        : I18n.t("forms.widgets.vlan.placeholderNoServicePort");

    const errorMessageExtra = missingInIms
        ? I18n.t("forms.widgets.vlan.missingInIms")
        : invalidFormat
        ? I18n.t("forms.widgets.vlan.invalidVlan")
        : !isEmpty(vlansInUse)
        ? vlansInUse.length >= 1 && vlansInUse[0] === 0
            ? I18n.t("forms.widgets.vlan.untaggedPortInUse")
            : I18n.t("forms.widgets.vlan.vlansInUseError", { vlans: vlansInUse.join(", ") })
        : undefined;
    const message = isEmpty(allUsedVlans)
        ? portMode === "tagged"
            ? I18n.t("forms.widgets.vlan.allPortsAvailable")
            : I18n.t("forms.widgets.vlan.taggedOnly")
        : I18n.t("forms.widgets.vlan.vlansInUse", { vlans: allUsedVlans });

    return (
        <section {...filterDOMProps(props)}>
            {label && (
                <label htmlFor={id}>
                    {label}
                    {description && <em>{description}</em>}
                </label>
            )}
            <input
                disabled={disabled || portMode !== "tagged"}
                id={id}
                name={name}
                onChange={event => onChange(event.target.value)}
                placeholder={placeholder}
                ref={inputRef}
                type={type}
                value={portMode !== "tagged" ? 0 : value}
                onBlur={e => {
                    const target = e.target as HTMLInputElement;
                    validateUsedVlans(target.value);
                }}
            ></input>
            {subscriptionId && !disabled && <em>{message}</em>}
            {error && showInlineError && (
                <em className="error">
                    <div className="backend-validation">{errorMessage || errorMessageExtra}</div>
                </em>
            )}
        </section>
    );
}

export default connectField(Vlan, { kind: "leaf" });
