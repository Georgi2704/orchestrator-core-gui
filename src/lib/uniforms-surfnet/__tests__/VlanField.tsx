import fetchMock from "fetch-mock";
import React from "react";

import VlanField from "../src/VlanField";
import createContext from "./_createContext";
import mount from "./_mount";

describe.skip("<VlanField>", () => {
    test.skip("", () => {});
});
