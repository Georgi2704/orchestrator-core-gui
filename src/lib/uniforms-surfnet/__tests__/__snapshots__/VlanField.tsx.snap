// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VlanField> <VlanField> - adds vlans taken in same list as used vlans 1`] = `
"<ListField name=\\"x\\">
  <List name=\\"x\\" id=\\"uniforms-0000-000j\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Array], optional: false, label: 'X' }} fieldType={[Function: Array]} fields={[]} onChange={[Function]} value={[ { subscription_id: 'abc', y: '2,3' }, { subscription_id: 'abc', y: '3' }, { subscription_id: 'def', y: '3' } ]} required={true}>
    <section id=\\"uniforms-0000-000j\\" required={true} className=\\" list-field\\">
      <ul>
        <NestField name=\\"0\\" outerList={false}>
          <Nest name=\\"x.0\\" outerList={false} id=\\"uniforms-0000-000l\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Object], optional: true, label: 'X' }} fieldType={[Function: Object]} fields={[ 'y' ]} onChange={[Function]} value={{ subscription_id: 'abc', y: '2,3' }} required={false}>
            <div id=\\"uniforms-0000-000l\\" required={false}>
              <VlanField name=\\"y\\">
                <Vlan name=\\"x.0.y\\" id=\\"uniforms-0000-000m\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: String], optional: false, label: 'Y' }} fieldType={[Function: String]} fields={[]} onChange={[Function]} value=\\"2,3\\" required={true}>
                  <section required={true}>
                    <input disabled={false} id=\\"uniforms-0000-000m\\" name=\\"x.0.y\\" onChange={[Function: onChange]} placeholder=\\"Enter a valid VLAN range...\\" type=\\"text\\" value=\\"2,3\\" />
                    <em className=\\"info\\">
                      Already used VLAN ranges for this service port: 3
                    </em>
                  </section>
                </Vlan>
              </VlanField>
            </div>
          </Nest>
        </NestField>
        <NestField name=\\"1\\" outerList={false}>
          <Nest name=\\"x.1\\" outerList={false} id=\\"uniforms-0000-000n\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Object], optional: true, label: 'X' }} fieldType={[Function: Object]} fields={[ 'y' ]} onChange={[Function]} value={{ subscription_id: 'abc', y: '3' }} required={false}>
            <div id=\\"uniforms-0000-000n\\" required={false}>
              <VlanField name=\\"y\\">
                <Vlan name=\\"x.1.y\\" id=\\"uniforms-0000-000o\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: String], optional: false, label: 'Y' }} fieldType={[Function: String]} fields={[]} onChange={[Function]} value=\\"3\\" required={true}>
                  <section required={true}>
                    <input disabled={false} id=\\"uniforms-0000-000o\\" name=\\"x.1.y\\" onChange={[Function: onChange]} placeholder=\\"Enter a valid VLAN range...\\" type=\\"text\\" value=\\"3\\" />
                    <em className=\\"info\\">
                      Already used VLAN ranges for this service port: 2-3
                    </em>
                  </section>
                </Vlan>
              </VlanField>
            </div>
          </Nest>
        </NestField>
        <NestField name=\\"2\\" outerList={false}>
          <Nest name=\\"x.2\\" outerList={false} id=\\"uniforms-0000-000p\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Object], optional: true, label: 'X' }} fieldType={[Function: Object]} fields={[ 'y' ]} onChange={[Function]} value={{ subscription_id: 'def', y: '3' }} required={false}>
            <div id=\\"uniforms-0000-000p\\" required={false}>
              <VlanField name=\\"y\\">
                <Vlan name=\\"x.2.y\\" id=\\"uniforms-0000-000q\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: String], optional: false, label: 'Y' }} fieldType={[Function: String]} fields={[]} onChange={[Function]} value=\\"3\\" required={true}>
                  <section required={true}>
                    <input disabled={false} id=\\"uniforms-0000-000q\\" name=\\"x.2.y\\" onChange={[Function: onChange]} placeholder=\\"Enter a valid VLAN range...\\" type=\\"text\\" value=\\"3\\" />
                    <em className=\\"info\\">
                      This service port has no VLANs in use (yet).
                    </em>
                  </section>
                </Vlan>
              </VlanField>
            </div>
          </Nest>
        </NestField>
        <ListAddField initialCount={1} name=\\"$\\" disabled={false} outerList={false}>
          <ListAdd initialCount={1} name=\\"x.$\\" disabled={false} outerList={false} id=\\"uniforms-0000-000r\\" label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: Object], optional: true, label: 'X' }} fieldType={[Function: Object]} fields={[ 'y' ]} onChange={[Function]} value={{}} required={false}>
            <div className=\\"add-item\\" id=\\"uniforms-0000-000r\\" required={false} onClick={[Function: onClick]}>
              <i className=\\"fa fa-plus \\" />
              <label />
            </div>
          </ListAdd>
        </ListAddField>
      </ul>
    </section>
  </List>
</ListField>"
`;

exports[`<VlanField> <VlanField> - renders an input that is correctly enabled/disabled based on portMode (tagged with used Vlans) 1`] = `
"<VlanField name=\\"x\\">
  <Vlan name=\\"x\\" id=\\"uniforms-0000-000i\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: String], optional: false, label: 'X' }} fieldType={[Function: String]} fields={[]} onChange={[Function]} value={[undefined]} required={true}>
    <section required={true}>
      <input disabled={false} id=\\"uniforms-0000-000i\\" name=\\"x\\" onChange={[Function: onChange]} placeholder=\\"Enter a valid VLAN range...\\" type=\\"text\\" value=\\"\\" />
      <em className=\\"info\\">
        Already used VLAN ranges for this service port: 3-6,10
      </em>
    </section>
  </Vlan>
</VlanField>"
`;

exports[`<VlanField> <VlanField> - renders an input that is correctly enabled/disabled based on portMode (tagged) 1`] = `
"<VlanField name=\\"x\\">
  <Vlan name=\\"x\\" id=\\"uniforms-0000-000h\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: String], optional: false, label: 'X' }} fieldType={[Function: String]} fields={[]} onChange={[Function]} value={[undefined]} required={true}>
    <section required={true}>
      <input disabled={false} id=\\"uniforms-0000-000h\\" name=\\"x\\" onChange={[Function: onChange]} placeholder=\\"Enter a valid VLAN range...\\" type=\\"text\\" value=\\"\\" />
      <em className=\\"info\\">
        This service port has no VLANs in use (yet).
      </em>
    </section>
  </Vlan>
</VlanField>"
`;

exports[`<VlanField> <VlanField> - renders an input that is correctly enabled/disabled based on portMode (untagged with used vlans) 1`] = `
"<VlanField name=\\"x\\">
  <Vlan name=\\"x\\" id=\\"uniforms-0000-000g\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: String], optional: false, label: 'X' }} fieldType={[Function: String]} fields={[]} onChange={[Function]} value=\\"0\\" required={true}>
    <section required={true}>
      <input disabled={true} id=\\"uniforms-0000-000g\\" name=\\"x\\" onChange={[Function: onChange]} placeholder=\\"Enter a valid VLAN range...\\" type=\\"text\\" value=\\"0\\" />
      <em className=\\"info\\">
        VLAN is only relevant for SN7 MSP or SN8 SP in tagged mode, not for link_member or untagged ports.
      </em>
    </section>
  </Vlan>
</VlanField>"
`;

exports[`<VlanField> <VlanField> - renders an input that is correctly enabled/disabled based on portMode (untagged) 1`] = `
"<VlanField name=\\"x\\">
  <Vlan name=\\"x\\" id=\\"uniforms-0000-000f\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: String], optional: false, label: 'X' }} fieldType={[Function: String]} fields={[]} onChange={[Function]} value=\\"0\\" required={true}>
    <section required={true}>
      <input disabled={true} id=\\"uniforms-0000-000f\\" name=\\"x\\" onChange={[Function: onChange]} placeholder=\\"Enter a valid VLAN range...\\" type=\\"text\\" value=\\"0\\" />
      <em className=\\"info\\">
        VLAN is only relevant for SN7 MSP or SN8 SP in tagged mode, not for link_member or untagged ports.
      </em>
    </section>
  </Vlan>
</VlanField>"
`;
